improver/nbhood/nbhood.py

```diff
@@ -329,16 +332,79 @@ def _calculate_neighbourhood(
             data[area_sum == 0] = np.nan
             data = data.clip(min_val, max_val)
 
-        # Output type.
-        if issubclass(data.dtype.type, np.complexfloating):
-            data_dtype = np.complex64
-        else:
-            data_dtype = np.float32
-        data = data.astype(data_dtype)
-
         if self.re_mask:
             data = np.ma.masked_array(data, data_mask, copy=False)
 
+        return data.astype(out_data_dtype)
+
+    def _do_nbhood_sum(
```

Comment:
This method does not work as it stands for an annulus of 1 values, e.g.

```
[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 0. 0. 1. 1. 1.]
 [1. 1. 1. 0. 0. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]]
```
Consider this case with a radius of 3 grid cells. The half radius is rounded to 2, meaning the array passed to the `boxsum` method is the above with 1 edge row/column trimmed off of each side.

```
[[1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 0. 0. 1. 1.]
 [1. 1. 0. 0. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]]
```
The `boxsum` method asymmetrically pads this array to perform the summing, adding zeros around the edges:
```
[[0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 1. 1. 1. 1. 1. 0.]
 [0. 0. 1. 1. 1. 1. 1. 1. 0.]
 [0. 0. 1. 1. 0. 0. 1. 1. 0.]
 [0. 0. 1. 1. 0. 0. 1. 1. 0.]
 [0. 0. 1. 1. 1. 1. 1. 1. 0.]
 [0. 0. 1. 1. 1. 1. 1. 1. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]]
```
We end up with a square of spuriously low values inside the edges that have been trimmed. We get this:

![reworked](https://github.com/metoppv/improver/assets/28042362/427240a0-eca6-4780-a7f9-cfc2a2cd7a06)

when we should get this:

![original](https://github.com/metoppv/improver/assets/28042362/1a8f22d4-1a0f-4831-93fc-2dc3d4a5381f)


You will need to provide the value to use in the padding (1) in these cases to get this to behave as expected.

We also need a unit test to cover this case.

Reply:
My immediate thought is that this isn't the behaviour we see.  I believe that `boxsum` performs the padding with a pad value of `constant`, which means that the nearest available value is used. In the case described above, this would be 1 and the result of `boxsum` would be what we expect. My second thought is that the padding is irrelevant anyway because all the data in the padding region (which is where the differences are) are discarded. Thirdly, I definitely looked at the behaviour of padded arrays, so I think there must be a test for this somewhere.

Reply:
Not surprisingly, @bayliffe was right! I've added two tests to show the behaviour for the example documented here. Turning the optimising block off and on showed that the code was producing these erroneous results. The fix was to simply increase the `half_nb_size` by one.

Reply:
After a bit more thinking, I've changed this again so that the test has a much larger array, ensuring we are triggering the shrinking optimisation. I have changed the shrinking so that it interacts better with the padding. The shrinking will now keep enough rows/columns to cover the diameter of the neighbourhood, rather than the radius. This ensures that output points within the radius of a non-extreme value, can be accurately calculated. There is also a check to disable this optimisation if the minimum array + diameter box intersects the edge of the original data array so that we can continue to use the documented padding method.